core/src/qos_service.rs

```diff
@@ -170,6 +170,33 @@ impl QosService {
         (select_results, num_included)
     }
 
+    /// Update the transaction cost in the cost_tracker with the real cost for
+    /// transactions that were executed successfully;
+    /// Otherwise remove the cost from the cost tracker, therefore preventing cost_tracker
+    /// being inflated with unsuccessfully executed transactions.
+    pub fn update_or_remove_transaction_costs<'a>(
+        transaction_costs: impl Iterator<Item = &'a TransactionCost>,
+        transaction_qos_results: impl Iterator<Item = &'a transaction::Result<()>>,
+        retryable_transaction_indexes: &[usize],
+        bank: &Arc<Bank>,
+    ) {
+        let mut cost_tracker = bank.write_cost_tracker().unwrap();
+        transaction_costs
+            .zip(transaction_qos_results)
+            .enumerate()
+            .for_each(|(index, (tx_cost, qos_result))| {
+                if qos_result.is_ok() && retryable_transaction_indexes.contains(&index) {
```

Comment:
jstary pointed out that a tx can be unexecuted without ending up in retryable_transaction_indices. Needs a fix.